bin/prefix-vendor-namespace.php

```diff
@@ -45,92 +96,165 @@
-			$uses_change
-		);
+	if ( $prefix_namespace ) {
+		prefix_namespace( $contents, $package );
+	}
 
-		if ( ! empty( $direct_replacements[ $path ] ) ) {
-			foreach ( $direct_replacements[ $path ] as $replace ) {
-				$direct_change = 0;
-				$replace       = preg_quote( $replace, '#' );
-				$contents      = preg_replace(
-					"#({$replace})#m",
-					"{$new_namespace}\\\\\$1",
-					$contents,
-					-1,
-					$direct_change
-				);
-				if ( $direct_change ) {
-					$uses_change += $direct_change;
-				}
+	if ( $prefix_uses ) {
+		prefix_uses( $contents, $package );
+
+		if ( ! empty( $direct_replacements[ $package['package'] ] ) ) {
+			foreach ( $direct_replacements[ $package['package'] ] as $search ) {
+				prefix_string( $contents, $search );
 			}
 		}
-		if ( ! $namespace_change && $uses_change ) {
-			$file_notices[] = $file;
-		}
+	}
 
+	// Only overwrite file if the contents have changed.
+	if ( $content_hash !== md5( $contents ) ) {
 		file_put_contents( $file, $contents );
 	}
+}
 
-	// Update the namespace in the composer.json files, recursively finding all files named explicitly "composer.json".
-	$composer_files = get_dir_contents(
-		"{$vendor_dir}/{$path}",
-		'/' . preg_quote( DIRECTORY_SEPARATOR, '/') . 'composer.json$/'
-	);
+/**
+ * Prefix the namespace.
+ *
+ * @since x.x.x
+ *
+ * @param string $contents File contents.
+ * @param string $package  Package prefix configuration.
+ */
+function prefix_namespace( &$contents, $package ) {
+	global $namespace_prefix;
+
+	$quoted = preg_quote( $package['namespace'], '#' );
 
-	array_map(
-		function( $file ) use ( $namespace, $new_namespace ) {
-			return replace_in_json_file( $file, $namespace, $new_namespace );
-		},
-		$composer_files
+	// Match only the full namespace when strict is enabled.
+	if ( $package['strict'] ) {
+		$regex = "#^(\s*)(namespace)\s*({$quoted};)#m";
+	} else {
+		$regex = "#^(\s*)(namespace)\s*({$quoted}[\\\\|;])#m";
+	}
+
+	$contents = preg_replace(
+		$regex,
+		"\$1\$2 {$namespace_prefix}\\\\\$3",
+		$contents
 	);
 
-	// Update the namespace in vendor/composer/installed.json
-	// This file is used to generate the classmaps.
-	replace_in_json_file( "{$vendor_dir}/composer/installed.json", $namespace, $new_namespace );
+	if ( $package['strict'] && ! empty( $package['extra_namespaces'] ) ) {
+		foreach ( $package['extra_namespaces'] as $namespace ) {
+			prefix_namespace(
+				$contents,
+				[
+					'namespace' => $namespace,
+					'package'   => $package['package'],
+					'strict'    => true,
+				]
+			);
+		}
+	}
+}
 
-	// Remove file autoloads from vendor/composer/installed.json
-	remove_file_autoloads(
-		"{$vendor_dir}/composer/installed.json",
-		$composer_json['autoload']['files'] ?? [],
-		$path
+/**
+ * Prefix any import statements and direct uses of namespace.
+ *
+ * @since x.x.x
+ *
+ * @param string $contents File contents.
+ * @param string $package  Package prefix configuration.
+ */
+function prefix_uses( &$contents, $package ) {
```

Comment:
Same as in https://github.com/woocommerce/google-listings-and-ads/pull/1859/files#r1085080705